### ISSUE TITLE
OpenSSL Cipher missing methods

### ISSUE THREAD

**Issue by jwoertink** [crystal-lang/crystal#9681]
In Ruby, the `OpenSSL::Cipher` class has an [auth_data](https://ruby-doc.org/stdlib-2.7.0/libdoc/openssl/rdoc/OpenSSL/Cipher.html#method-i-auth_data-3D) and [auth_tag](https://ruby-doc.org/stdlib-2.7.0/libdoc/openssl/rdoc/OpenSSL/Cipher.html#method-i-auth_tag) methods that seem to be missing from our [Cipher](https://github.com/crystal-lang/crystal/blob/master/src/openssl/cipher.cr). 

I'm not really sure what these do, but I was trying to port some [code from Ruby](https://github.com/mdp/gibberish/blob/master/lib/gibberish/aes.rb#L142-L149) over, and got stuck on these methods.

**Comment by washu** [comment]
did you find a solution?

**Comment by jwoertink** [comment]
Not exactly. Ended up writing my own mini lib https://github.com/jwoertink/psst It doesn't really solve the issue, but allowed me to move on.

**Comment by washu** [comment]
i got it working with a weird caveat.
I monkeypatched lib_crypto to access the set data flag used in ruby
im my use case im reading the credentails.yml.enc file decrypting it to get the db username/passwords


here is the code im using
`         # key is loaded from the .enc file OR from a docker secrets file skipped for brevity 
          key = key.hexbytes
          # load the encrypted file
          encrypted = File.open(load_file) do |file|
            file.gets_to_end
          end
          # copied from rails message encryptor and translated to crystal
          encrypted_data, iv, auth_tag = encrypted.split("--").map { |v| Base64.decode(v) }
          cipher = OpenSSL::Cipher.new("aes-128-gcm")
          cipher.decrypt
          cipher.key = key
          cipher.iv = iv
          cipher.auth_tag = auth_tag
          io = IO::Memory.new
          begin
            # rubys method that does chiper.auth_data = "", under the hoods calls down the chipher update. so i do it here first
            io.write(cipher.update(""))
            io.write(cipher.update(encrypted_data))
            io.write(cipher.final)
          rescue e : Exception
            puts "Oddly we fail decryption here it could be the message auth tag is invalid or not set in the underlying ssl lib"
          end
          io.rewind
          decrypted_data = io.gets_to_end
          # for some reason we get a BOM marker as first 4 bytes and last byte is always an extra \" so we trim it off
          return YAML.parse(decrypted_data[5..-1])
`

I cant ever get the final to not error, but even not calling final the data is decrypted not exactly sure why that is but it works

here is the monkeypatch
`
@[Link("crypto")]
lib LibCrypto
	EVP_CTRL_GCM_GET_TAG = 0x10
	EVP_CTRL_GCM_SET_TAG = 0x11
	EVP_GCM_TLS_TAG_LEN  = 16
	fun evp_cipher_ctx_ctrl = EVP_CIPHER_CTX_ctrl(ctx : EVP_CIPHER_CTX, type : Int32, arg : Int32, ptr : Void*) : Int32
end
class OpenSSL::Cipher
	def auth_tag() : Bytes
		tag = Bytes.new(LibCrypto::EVP_GCM_TLS_TAG_LEN)
		LibCrypto.evp_cipher_ctx_ctrl(@ctx, LibCrypto::EVP_CTRL_GCM_GET_TAG, tag.size, pointerof(tag).as(Void*))
		return tag
	end
	def auth_tag=(tag : Bytes)
		LibCrypto.evp_cipher_ctx_ctrl(@ctx, LibCrypto::EVP_CTRL_GCM_SET_TAG, tag.size, pointerof(tag).as(Void*))
	end
end
`

**Comment by miry** [comment]
I have a bit modified the suggestion from @washu 

```crystal
# ext/std/lib_crypto.cr
@[Link("crypto")]
lib LibCrypto
  EVP_CTRL_GCM_GET_TAG  = 0x10
  EVP_CTRL_GCM_SET_TAG  = 0x11
  EVP_CTRL_AEAD_GET_TAG = 0x10
  EVP_CTRL_AEAD_SET_TAG = 0x11
  EVP_GCM_TLS_TAG_LEN   =   16

  # https://docs.openssl.org/3.3/man3/EVP_EncryptInit/#synopsis
  # int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int cmd, int p1, void *p2);
  fun evp_cipher_ctx_ctrl = EVP_CIPHER_CTX_ctrl(ctx : EVP_CIPHER_CTX, cmd : Int32, p1 : Int32, ptr : Void*) : Int32
end

class OpenSSL::Cipher
  def auth_tag(tag_len = LibCrypto::EVP_GCM_TLS_TAG_LEN) : Bytes
    if !authenticated?
      raise Error.new "EVP_CipherAuthTag: authentication tag not supported by this cipher"
    end
    tag = Bytes.new(tag_len)
    res = LibCrypto.evp_cipher_ctx_ctrl(@ctx, LibCrypto::EVP_CTRL_AEAD_GET_TAG, tag_len, tag)
    if res != 1
      raise Error.new "EVP_CipherCtxCtrl: retrieving the authentication tag failed"
    end
    tag
  end

  def auth_tag=(tag : Bytes)
    if !authenticated?
      raise Error.new "EVP_CipherAuthTag: authentication tag not supported by this cipher"
    end
    res = LibCrypto.evp_cipher_ctx_ctrl(@ctx, LibCrypto::EVP_CTRL_AEAD_SET_TAG, tag.size, tag)
    if res != 1
      raise Error.new "EVP_CipherCtxCtrl: unable to set AEAD tag"
    end
    tag
  end
end
```

Add a bit of testings:

```crystal
class OpenSSL::CipherTest < Minitest::Test
  def test_auth_tag
    cipher = OpenSSL::Cipher.new("aes-256-gcm")
    assert_equal "id-aes256-GCM", cipher.name
    assert cipher.authenticated?

    cipher.encrypt
    cipher.key = "1" * 32
    # deterministic
    cipher.iv = "1" * 12

    encrypted_data : Bytes = cipher.update("foo")
    encrypted_data += cipher.final

    assert_equal Bytes[16, 57, 248], encrypted_data
    assert_equal Bytes[105, 217, 223, 177, 218, 85, 94, 175, 92, 140, 224, 0, 117, 9, 59, 195], cipher.auth_tag
  end
end
```

I am also working on porting Rails implementation of ActiveRecord::Encryption and it seems this solution is working.

**Comment by washu** [comment]
be aware the encrypted file when decoded, has marshalling data so io.pos = 6 to skip past and them and json decoded (if using default rails setup), the rails json parse just skips those bytes in the json gem as a parsing option
